<a href="https://colab.research.google.com/github/koalalovepabro/Stock-Portfolio-Optimization/blob/main/Build_A_Killer_Stock_Portfolio(KRX).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 최적화된 주식 포트폴리오 만들기

## 💘 한국 주식 적용하기

In [1]:
!pip install -q install finance-datareader

In [2]:
import pandas as pd
import numpy as np
import requests

In [3]:
# Finance Datareader 불러오고, 버전 확인하기
import FinanceDataReader as fdr
fdr.__version__

'0.9.31'

In [4]:
# KRX : KRX 종목 전체
# KOSPI : KOSPI 종목
# KOSDAQ : KOSDAQ 종목
# KONEX : KONEX 종목
# NASDAQ : 나스닥 종목
# NYSE : 뉴욕증권거래소 종목
# SP500 : S&P500 종목

In [5]:
# 한국거래소(KRX) 상장종목 전체 가져오기
df_krx = fdr.StockListing('KRX')
df_krx

Symbol  Market  ...                       HomePage Region
0     060310  KOSDAQ  ...           http://www.3sref.com  서울특별시
1     095570   KOSPI  ...         http://www.ajnet.co.kr  서울특별시
2     006840   KOSPI  ...  http://www.aekyunggroup.co.kr  서울특별시
3     054620  KOSDAQ  ...   http://www.apsholdings.co.kr    경기도
4     265520  KOSDAQ  ...     http://www.apsystems.co.kr    경기도
...      ...     ...  ...                            ...    ...
6460  000547   KOSPI  ...                            NaN    NaN
6461  000545   KOSPI  ...                            NaN    NaN
6462  003280   KOSPI  ...         http://www.heung-a.com  서울특별시
6463  037440  KOSDAQ  ...          http://www.heerim.com  서울특별시
6464  238490  KOSDAQ  ...          http://www.hims.co.kr  인천광역시

[6465 rows x 10 columns]

In [6]:
# 데이터 파악
# 데이터의 결측치 존재 여부, 데이터의 타입, 컬럼의 수, 데이터프레임의 메모리 사용량 등
df_krx.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 6465 entries, 0 to 6464
Data columns (total 10 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   Symbol          6465 non-null   object        
 1   Market          6465 non-null   object        
 2   Name            6465 non-null   object        
 3   Sector          2450 non-null   object        
 4   Industry        2431 non-null   object        
 5   ListingDate     2450 non-null   datetime64[ns]
 6   SettleMonth     2450 non-null   object        
 7   Representative  2450 non-null   object        
 8   HomePage        2269 non-null   object        
 9   Region          2450 non-null   object        
dtypes: datetime64[ns](1), object(9)
memory usage: 555.6+ KB


In [7]:
# 데이터의 결측치 확인
df_krx.isnull().sum()

Symbol               0
Market               0
Name                 0
Sector            4015
Industry          4034
ListingDate       4015
SettleMonth       4015
Representative    4015
HomePage          4196
Region            4015
dtype: int64

In [8]:
# 결측치 제거
df_krx = df_krx.dropna()
df_krx.isnull().sum()

Symbol            0
Market            0
Name              0
Sector            0
Industry          0
ListingDate       0
SettleMonth       0
Representative    0
HomePage          0
Region            0
dtype: int64

In [9]:
df_krx.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2256 entries, 0 to 6464
Data columns (total 10 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   Symbol          2256 non-null   object        
 1   Market          2256 non-null   object        
 2   Name            2256 non-null   object        
 3   Sector          2256 non-null   object        
 4   Industry        2256 non-null   object        
 5   ListingDate     2256 non-null   datetime64[ns]
 6   SettleMonth     2256 non-null   object        
 7   Representative  2256 non-null   object        
 8   HomePage        2256 non-null   object        
 9   Region          2256 non-null   object        
dtypes: datetime64[ns](1), object(9)
memory usage: 193.9+ KB


In [10]:
# 데이터프레임의 통계정보
df_krx.describe()

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: FutureWarning: Treating datetime data as categorical rather than numeric in `.describe` is deprecated and will be removed in a future version of pandas. Specify `datetime_is_numeric=True` to silence this warning and adopt the future behavior now.
  


Symbol  Market   Name  ... Representative                 HomePage Region
count     2256    2256   2256  ...           2256                     2256   2256
unique    2256       3   2256  ...           2158                     2237     22
top     237690  KOSDAQ  한세엠케이  ...              -  http://www.moorim.co.kr  서울특별시
freq         1    1352      1  ...              5                        2    888
first      NaN     NaN    NaN  ...            NaN                      NaN    NaN
last       NaN     NaN    NaN  ...            NaN                      NaN    NaN

[6 rows x 10 columns]

In [11]:
# 종목코드 가져오기
assets = df_krx['Symbol']
assets

0       060310
1       095570
2       006840
3       054620
4       265520
         ...  
6458    189980
6459    000540
6462    003280
6463    037440
6464    238490
Name: Symbol, Length: 2256, dtype: object

In [12]:
assets = np.array(assets)
assets

array(['060310', '095570', '006840', ..., '003280', '037440', '238490'],
      dtype=object)

In [13]:
len(assets)

2256

In [16]:
# 종목별 종가 가져오기
from datetime import datetime

# 주식 시작일은 2013년 1월 1일이고 종료일은 현재 날짜 (오늘)로 설정
#Get the stock starting date
start_date = '2013-01-01'
# today = datetime.today().strftime('%Y-%m-%d')
end_date = '2021-07-15'

# 각 주식의 일별 종가 데이터를 저장할 데이터 프레임을 생성
#Create a dataframe to store the adjusted close price of the stocks
df = pd.DataFrame()

# FinanceDataReader로 각 종목의 종가데이터 불러오기
for stock in assets:
  df[stock] = fdr.DataReader(stock, start_date, end_date)['Close']
df

060310  095570  006840  054620  ...  000540  003280  037440   238490
Date                                        ...                                 
2013-01-02    7150     NaN   21565    5722  ...    4915    6352    6730      NaN
2013-01-03    7090     NaN   22292    6035  ...    4965    6345    6770      NaN
2013-01-04    7330     NaN   22971    6296  ...    4900    6572    6700      NaN
2013-01-07    7850     NaN   23068    6236  ...    4905    6696    6680      NaN
2013-01-08    8960     NaN   22971    6236  ...    4910    6696    6720      NaN
...            ...     ...     ...     ...  ...     ...     ...     ...      ...
2021-07-09    2955  5850.0   27650   14850  ...    4340    1035    7420  10950.0
2021-07-12    2940  5910.0   28550   14850  ...    4395    1035    7550  11100.0
2021-07-13    2930  5920.0   29050   15700  ...    4420    1035    7360  11400.0
2021-07-14    2900  6330.0   28500   17850  ...    4425    1035    7150  11350.0
2021-07-15    2935  6200.0   29100   18000  ...    4415    1035    7170  11450.0

[2101 rows x 2256 columns]

In [17]:
df

060310  095570  006840  054620  ...  000540  003280  037440   238490
Date                                        ...                                 
2013-01-02    7150     NaN   21565    5722  ...    4915    6352    6730      NaN
2013-01-03    7090     NaN   22292    6035  ...    4965    6345    6770      NaN
2013-01-04    7330     NaN   22971    6296  ...    4900    6572    6700      NaN
2013-01-07    7850     NaN   23068    6236  ...    4905    6696    6680      NaN
2013-01-08    8960     NaN   22971    6236  ...    4910    6696    6720      NaN
...            ...     ...     ...     ...  ...     ...     ...     ...      ...
2021-07-09    2955  5850.0   27650   14850  ...    4340    1035    7420  10950.0
2021-07-12    2940  5910.0   28550   14850  ...    4395    1035    7550  11100.0
2021-07-13    2930  5920.0   29050   15700  ...    4420    1035    7360  11400.0
2021-07-14    2900  6330.0   28500   17850  ...    4425    1035    7150  11350.0
2021-07-15    2935  6200.0   29100   18000  ...    4415    1035    7170  11450.0

[2101 rows x 2256 columns]

In [22]:
# DataFrame을 csv 파일로 저장하기 ( 결측값 제거하지 않음 )
df.to_csv("krx_code_close.csv", index=True)

In [29]:
# 칼럼명을 회사이름으로 변경
df.columns = df_krx['Name'].values
df

3S  AJ네트웍스  AK홀딩스  APS홀딩스  ...  흥국화재  흥아해운    희림       힘스
Date                                     ...                           
2013-01-02  7150     NaN  21565    5722  ...  4915  6352  6730      NaN
2013-01-03  7090     NaN  22292    6035  ...  4965  6345  6770      NaN
2013-01-04  7330     NaN  22971    6296  ...  4900  6572  6700      NaN
2013-01-07  7850     NaN  23068    6236  ...  4905  6696  6680      NaN
2013-01-08  8960     NaN  22971    6236  ...  4910  6696  6720      NaN
...          ...     ...    ...     ...  ...   ...   ...   ...      ...
2021-07-09  2955  5850.0  27650   14850  ...  4340  1035  7420  10950.0
2021-07-12  2940  5910.0  28550   14850  ...  4395  1035  7550  11100.0
2021-07-13  2930  5920.0  29050   15700  ...  4420  1035  7360  11400.0
2021-07-14  2900  6330.0  28500   17850  ...  4425  1035  7150  11350.0
2021-07-15  2935  6200.0  29100   18000  ...  4415  1035  7170  11450.0

[2101 rows x 2256 columns]

In [30]:
# 결측값 있는 열 삭제  ( 종목 2256 -> 1476으로 줄어 듦 )
df2 = df.dropna(axis = 1)
df2

3S  AK홀딩스  APS홀딩스  BNK금융지주  ...     흥국  흥국화재  흥아해운    희림
Date                                      ...                         
2013-01-02  7150  21565    5722    12536  ...   1928  4915  6352  6730
2013-01-03  7090  22292    6035    12816  ...   1925  4965  6345  6770
2013-01-04  7330  22971    6296    12955  ...   1925  4900  6572  6700
2013-01-07  7850  23068    6236    13049  ...   1915  4905  6696  6680
2013-01-08  8960  22971    6236    13142  ...   1923  4910  6696  6720
...          ...    ...     ...      ...  ...    ...   ...   ...   ...
2021-07-09  2955  27650   14850     7350  ...  11850  4340  1035  7420
2021-07-12  2940  28550   14850     7520  ...  11700  4395  1035  7550
2021-07-13  2930  29050   15700     7590  ...  11550  4420  1035  7360
2021-07-14  2900  28500   17850     7520  ...  11650  4425  1035  7150
2021-07-15  2935  29100   18000     7560  ...  11550  4415  1035  7170

[2101 rows x 1476 columns]

In [32]:
df2.isnull().sum()

3S         0
AK홀딩스      0
APS홀딩스     0
BNK금융지주    0
BYC        0
          ..
휴켐스        0
흥국         0
흥국화재       0
흥아해운       0
희림         0
Length: 1476, dtype: int64

In [33]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 2101 entries, 2013-01-02 to 2021-07-15
Columns: 1476 entries, 3S to 희림
dtypes: int64(1476)
memory usage: 23.7 MB


In [34]:
# 결측값을 가진 열을 제거한 DataFrame을 csv 파일로 저장하기
df2.to_csv("krx_name_close_drop_columns.csv", index=True)

In [35]:
# Get the assets / tickers
assets = df2.columns
assets

Index(['3S', 'AK홀딩스', 'APS홀딩스', 'BNK금융지주', 'BYC', 'CBI', 'CJ', 'CJ CGV',
       'CJ ENM', 'CJ대한통운',
       ...
       '휴먼엔', '휴비스', '휴비츠', '휴스틸', '휴온스글로벌', '휴켐스', '흥국', '흥국화재', '흥아해운',
       '희림'],
      dtype='object', length=1476)

In [36]:
len(assets)

1476

In [37]:
df2.isnull().sum()

3S         0
AK홀딩스      0
APS홀딩스     0
BNK금융지주    0
BYC        0
          ..
휴켐스        0
흥국         0
흥국화재       0
흥아해운       0
희림         0
Length: 1476, dtype: int64

- Index : 날짜 (2,101  days)
- column : 종목명 (1,476개 종목)
- row : 날짜별 종목의 종가(Close Price)

In [38]:
!pip install PyPortfolioOpt

     |████████████████████████████████| 61kB 4.6MB/s 
     |████████████████████████████████| 1.3MB 17.8MB/s 
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
    Preparing wheel metadata ... done
  Created wheel for cvxpy: filename=cvxpy-1.1.13-cp37-cp37m-linux_x86_64.whl size=2736188 sha256=0daddd47c5be6ecee81a009072aa44cb44d55d796ead1d342cbb81544b82cda9
  Stored in directory: /root/.cache/pip/wheels/f9/78/00/f29636789ee83434953b5442f16ec3f9834a68e7fd0393c220
Successfully built cvxpy
  Found existing installation: cvxpy 1.0.31
    Uninstalling cvxpy-1.0.31:
      Successfully uninstalled cvxpy-1.0.31


In [39]:
# Optimize the portfolio
from pypfopt.efficient_frontier import EfficientFrontier
from pypfopt import risk_models
from pypfopt import expected_returns

In [40]:
# Calculate the expected annualized returns and the annualized sample covariance matrix of the daily asset returns
mu = expected_returns.mean_historical_return(df2)
S = risk_models.sample_cov(df2)

In [41]:
mu

3S        -0.101338
AK홀딩스      0.036614
APS홀딩스     0.147432
BNK금융지주   -0.058883
BYC        0.096404
             ...   
휴켐스       -0.008212
흥국         0.239644
흥국화재      -0.012792
흥아해운      -0.195653
희림         0.007629
Length: 1476, dtype: float64

In [42]:
S

3S     AK홀딩스    APS홀딩스  ...      흥국화재      흥아해운        희림
3S       0.299297  0.045523  0.049292  ...  0.029733  0.040025  0.029341
AK홀딩스    0.045523  0.180536  0.044027  ...  0.036007  0.048133  0.017887
APS홀딩스   0.049292  0.044027  0.274046  ...  0.037520  0.040505  0.016974
BNK금융지주  0.016259  0.028559  0.018321  ...  0.020180  0.017324  0.006361
BYC      0.020442  0.013711  0.022236  ...  0.017223  0.014118  0.007938
...           ...       ...       ...  ...       ...       ...       ...
휴켐스      0.027589  0.031881  0.029540  ...  0.020934  0.014736  0.009909
흥국       0.047070  0.037078  0.043052  ...  0.034823  0.034614  0.026444
흥국화재     0.029733  0.036007  0.037520  ...  0.118429  0.019543  0.020995
흥아해운     0.040025  0.048133  0.040505  ...  0.019543  0.350154  0.019390
희림       0.029341  0.017887  0.016974  ...  0.020995  0.019390  0.133327

[1476 rows x 1476 columns]

In [43]:
S.shape

(1476, 1476)

In [49]:
!pip install cvxpy

In [55]:
# Optimize for the maximal Sharpe ratio
# 💛데이터셋이 너무 많으면, ef.max_sharpe()에서 에러남 -> solver를 SCS로 바꿔줌
# Rober says: 100개 이하로 종목을 추린 후에 실행시키기를 추천함 !
ef = EfficientFrontier(mu, S, solver="SCS")  # Create the Efficient Frontier Object

# Maximize the Sharpe ratio, and get the raw weights
weights = ef.max_sharpe()
cleaned_weights = ef.clean_weights()
print(cleaned_weights)

ef.portfolio_performance(verbose=True)

OrderedDict([('3S', 0.0), ('AK홀딩스', 0.0), ('APS홀딩스', 0.0), ('BNK금융지주', 0.0), ('BYC', 0.0), ('CBI', 0.0), ('CJ', 0.0), ('CJ CGV', 0.0), ('CJ ENM', 0.0), ('CJ대한통운', 0.0), ('CJ씨푸드', 0.0), ('CJ제일제당', 0.0), ('CJ프레시웨이', 0.0), ('CMG제약', 0.0), ('CNH', 0.0), ('CNT85', 0.0), ('COWON', 0.0), ('CS', 0.0), ('CSA 코스믹', 0.0), ('CS홀딩스', 0.0), ('DB', 0.0), ('DB금융투자', 0.0), ('DB손해보험', 0.0), ('DB하이텍', 0.0), ('DGB금융지주', 0.0), ('DI동일', 0.0), ('DL', 0.0), ('DL건설', 0.01595), ('DMS', 0.0), ('DRB동일', 0.0), ('DSR제강', 0.0), ('E1', 0.0), ('EG', 0.0), ('EMW', 0.0), ('ES큐브', 0.0), ('F&F홀딩스', 0.01117), ('GKL', 0.0), ('GST', 0.0), ('GS건설', 0.0), ('GS글로벌', 0.0), ('GS리테일', 0.0), ('GV', 0.0), ('HB테크놀러지', 0.0), ('HDC', 0.0), ('HDC현대EP', 0.0), ('HMM', 0.0), ('HRS', 0.0), ('HSD엔진', 0.0), ('IHQ', 0.0), ('ISC', 0.0), ('ITX-AI', 0.0), ('JW신약', 0.0), ('JW중외제약', 0.0), ('JW홀딩스', 0.0), ('JYP Ent.', 0.0), ('KBI메탈', 0.0), ('KB금융', 0.0), ('KB오토시스', 0.0), ('KCC', 0.0), ('KCC건설', 0.0), ('KCI', 0.0), ('KCTC', 0.0), ('KC그린홀딩스', 0.0), ('

(0.28876802253445577, 0.12040463155720918, 2.2322066772552103)

In [56]:
weights

OrderedDict([('3S', 6.67702695e-08),
             ('AK홀딩스', -5.05510866e-08),
             ('APS홀딩스', -4.073610503e-07),
             ('BNK금융지주', 9.193961781e-07),
             ('BYC', 3.750789046e-07),
             ('CBI', 4.662768522e-07),
             ('CJ', 1.784672276e-07),
             ('CJ CGV', 1.3996549353e-06),
             ('CJ ENM', -8.736482439e-07),
             ('CJ대한통운', -1.524856696e-06),
             ('CJ씨푸드', 1.0293363634e-06),
             ('CJ제일제당', -2.14504981e-07),
             ('CJ프레시웨이', 1.189814069e-07),
             ('CMG제약', -1.274605619e-07),
             ('CNH', -6.810159263e-07),
             ('CNT85', 9.1246024e-09),
             ('COWON', 7.77433994e-08),
             ('CS', 5.29007312e-08),
             ('CSA 코스믹', -5.563413925e-07),
             ('CS홀딩스', 8.943831257e-07),
             ('DB', 1.626011034e-07),
             ('DB금융투자', 4.632684237e-07),
             ('DB손해보험', 1.8070076927e-06),
             ('DB하이텍', 7.49063975e-07),
             ('DGB

In [57]:
cleaned_weights

OrderedDict([('3S', 0.0),
             ('AK홀딩스', 0.0),
             ('APS홀딩스', 0.0),
             ('BNK금융지주', 0.0),
             ('BYC', 0.0),
             ('CBI', 0.0),
             ('CJ', 0.0),
             ('CJ CGV', 0.0),
             ('CJ ENM', 0.0),
             ('CJ대한통운', 0.0),
             ('CJ씨푸드', 0.0),
             ('CJ제일제당', 0.0),
             ('CJ프레시웨이', 0.0),
             ('CMG제약', 0.0),
             ('CNH', 0.0),
             ('CNT85', 0.0),
             ('COWON', 0.0),
             ('CS', 0.0),
             ('CSA 코스믹', 0.0),
             ('CS홀딩스', 0.0),
             ('DB', 0.0),
             ('DB금융투자', 0.0),
             ('DB손해보험', 0.0),
             ('DB하이텍', 0.0),
             ('DGB금융지주', 0.0),
             ('DI동일', 0.0),
             ('DL', 0.0),
             ('DL건설', 0.01595),
             ('DMS', 0.0),
             ('DRB동일', 0.0),
             ('DSR제강', 0.0),
             ('E1', 0.0),
             ('EG', 0.0),
             ('EMW', 0.0),
             ('ES큐브', 0.0),


In [58]:
!pip install pulp

     |████████████████████████████████| 40.6MB 74kB/s 


In [81]:
# Get the discrete allocation of each sharpe per stock
from pypfopt.discrete_allocation import DiscreteAllocation, get_latest_prices

# 투자금액 (단위: KRW)
portfolio_val = 5000000

latest_prices = get_latest_prices(df2)

weights = cleaned_weights

da = DiscreteAllocation(weights, latest_prices, total_portfolio_value=portfolio_val)

allocation, leftover = da.lp_portfolio()

print('Discrete Allocaion: ', allocation)
print('Funds Remaining: ', leftover, ' KRW')

Discrete Allocaion:  {'DL건설': 2, 'F&F홀딩스': 2, 'MH에탄올': 1, 'NICE평가정보': 2, '고려신용정보': 31, '국일제지': 36, '까뮤이앤씨': 25, '나이스정보통신': 2, '네이처셀': 4, '대성홀딩스': 10, '대한방직': 2, '동남합성': 4, '동원개발': 3, '동화기업': 1, '리노공업': 1, '맥쿼리인프라': 45, '메리츠금융지주': 5, '미원상사': 3, '부산가스': 1, '빅텍': 12, '삼성공조': 1, '쌍용C&E': 2, '아난티': 3, '에이치엘비': 1, '에이치엘비생명과학': 3, '에이텍': 2, '우리기술투자': 5, '이크레더블': 4, '지어소프트': 1, '진원생명과학': 1, '카카오': 1, '티씨케이': 1, '파라텍': 3, '포스코강판': 1, '푸른저축은행': 10, '한국가구': 6, '한국기업평가': 3, '한솔케미칼': 1}
Funds Remaining:  685.0  KRW


In [82]:
allocation

{'DL건설': 2,
 'F&F홀딩스': 2,
 'MH에탄올': 1,
 'NICE평가정보': 2,
 '고려신용정보': 31,
 '국일제지': 36,
 '까뮤이앤씨': 25,
 '나이스정보통신': 2,
 '네이처셀': 4,
 '대성홀딩스': 10,
 '대한방직': 2,
 '동남합성': 4,
 '동원개발': 3,
 '동화기업': 1,
 '리노공업': 1,
 '맥쿼리인프라': 45,
 '메리츠금융지주': 5,
 '미원상사': 3,
 '부산가스': 1,
 '빅텍': 12,
 '삼성공조': 1,
 '쌍용C&E': 2,
 '아난티': 3,
 '에이치엘비': 1,
 '에이치엘비생명과학': 3,
 '에이텍': 2,
 '우리기술투자': 5,
 '이크레더블': 4,
 '지어소프트': 1,
 '진원생명과학': 1,
 '카카오': 1,
 '티씨케이': 1,
 '파라텍': 3,
 '포스코강판': 1,
 '푸른저축은행': 10,
 '한국가구': 6,
 '한국기업평가': 3,
 '한솔케미칼': 1}

In [83]:
len(allocation)

38

In [84]:
# 포트폴리오에 포함된 종목을 리스트로 만들기
company_name = list(allocation)

In [85]:
company_name

['DL건설',
 'F&F홀딩스',
 'MH에탄올',
 'NICE평가정보',
 '고려신용정보',
 '국일제지',
 '까뮤이앤씨',
 '나이스정보통신',
 '네이처셀',
 '대성홀딩스',
 '대한방직',
 '동남합성',
 '동원개발',
 '동화기업',
 '리노공업',
 '맥쿼리인프라',
 '메리츠금융지주',
 '미원상사',
 '부산가스',
 '빅텍',
 '삼성공조',
 '쌍용C&E',
 '아난티',
 '에이치엘비',
 '에이치엘비생명과학',
 '에이텍',
 '우리기술투자',
 '이크레더블',
 '지어소프트',
 '진원생명과학',
 '카카오',
 '티씨케이',
 '파라텍',
 '포스코강판',
 '푸른저축은행',
 '한국가구',
 '한국기업평가',
 '한솔케미칼']

In [86]:
len(company_name)

38

In [87]:
# Get the discrete allocation values
discrete_allocation_list = []
for symbol in allocation:
  discrete_allocation_list.append(allocation.get(symbol))

In [88]:
discrete_allocation_list

[2,
 2,
 1,
 2,
 31,
 36,
 25,
 2,
 4,
 10,
 2,
 4,
 3,
 1,
 1,
 45,
 5,
 3,
 1,
 12,
 1,
 2,
 3,
 1,
 3,
 2,
 5,
 4,
 1,
 1,
 1,
 1,
 3,
 1,
 10,
 6,
 3,
 1]

In [89]:
# Create a dataframe for the portfolio
portfolio_df = pd.DataFrame(columns = ['Company_name', 'company_Ticker', 'Discrete_val_'+str(portfolio_val)])
portfolio_df

Empty DataFrame
Columns: [Company_name, company_Ticker, Discrete_val_5000000]
Index: []

In [90]:
portfolio_df['Company_name'] = company_name
portfolio_df['company_Ticker'] = allocation
portfolio_df['Discrete_val_'+str(portfolio_val)] = discrete_allocation_list

In [91]:
# Show the portfolio
portfolio_df

Company_name company_Ticker  Discrete_val_5000000
0          DL건설           DL건설                     2
1        F&F홀딩스         F&F홀딩스                     2
2         MH에탄올          MH에탄올                     1
3      NICE평가정보       NICE평가정보                     2
4        고려신용정보         고려신용정보                    31
5          국일제지           국일제지                    36
6         까뮤이앤씨          까뮤이앤씨                    25
7       나이스정보통신        나이스정보통신                     2
8          네이처셀           네이처셀                     4
9         대성홀딩스          대성홀딩스                    10
10         대한방직           대한방직                     2
11         동남합성           동남합성                     4
12         동원개발           동원개발                     3
13         동화기업           동화기업                     1
14         리노공업           리노공업                     1
15       맥쿼리인프라         맥쿼리인프라                    45
16      메리츠금융지주        메리츠금융지주                     5
17         미원상사           미원상사                     3
18         부산가스           부산가스                     1
19           빅텍             빅텍                    12
20         삼성공조           삼성공조                     1
21        쌍용C&E          쌍용C&E                     2
22          아난티            아난티                     3
23        에이치엘비          에이치엘비                     1
24    에이치엘비생명과학      에이치엘비생명과학                     3
25          에이텍            에이텍                     2
26       우리기술투자         우리기술투자                     5
27        이크레더블          이크레더블                     4
28        지어소프트          지어소프트                     1
29       진원생명과학         진원생명과학                     1
30          카카오            카카오                     1
31         티씨케이           티씨케이                     1
32          파라텍            파라텍                     3
33        포스코강판          포스코강판                     1
34       푸른저축은행         푸른저축은행                    10
35         한국가구           한국가구                     6
36       한국기업평가         한국기업평가                     3
37        한솔케미칼          한솔케미칼                     1

In [92]:
# Show Funds Remaining
print('Funds Remaining: ', leftover, ' KRW')

# Show Portfolio performance
ef.portfolio_performance(verbose=True)

Funds Remaining:  685.0  KRW
Expected annual return: 28.9%
Annual volatility: 12.0%
Sharpe Ratio: 2.23


(0.28876802253445577, 0.12040463155720918, 2.2322066772552103)